File: VitalFace/src/VitalFace.Web/Services/KioskInterop.cs
using Microsoft.JSInterop;
using VitalFace.Web.Models;

namespace VitalFace.Web.Services;

/// <summary>
/// Typed wrapper around the wwwroot/js/kiosk/kiosk-interop.js ES module. Kept as the single seam
/// between Razor components and JS, so the kiosk page never calls IJSRuntime directly.
/// </summary>
public sealed class KioskInterop(IJSRuntime jsRuntime) : IAsyncDisposable
{
    private readonly Lazy<Task<IJSObjectReference>> _moduleTask = new(() =>
        jsRuntime.InvokeAsync<IJSObjectReference>("import", "./js/kiosk/kiosk-interop.js").AsTask());

    public async Task InitializeCaptureStageAsync(string videoContainerId, string vizContainerId)
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("initializeCaptureStage", videoContainerId, vizContainerId);
    }

    public async Task StartCaptureAsync()
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("startCapture");
    }

    public async Task<CaptureBuffers?> PeekBuffersAsync()
    {
        var module = await _moduleTask.Value;
        return await module.InvokeAsync<CaptureBuffers?>("peekBuffers");
    }

    public async Task<CaptureBuffers> StopCaptureAsync()
    {
        var module = await _moduleTask.Value;
        return await module.InvokeAsync<CaptureBuffers>("stopCapture");
    }

    public async Task UpdateLiveMetricsAsync(double bpm, double respPhase, double perclos, double confidence)
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("updateLiveMetrics", bpm, respPhase, perclos, confidence);
    }

    public async Task InitializeConsentAsync(string containerId)
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("initializeConsent", containerId);
    }

    public async Task ClearSignatureAsync()
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("clearSignature");
    }

    public async Task<bool> HasConsentSignatureAsync()
    {
        var module = await _moduleTask.Value;
        return await module.InvokeAsync<bool>("hasConsentSignature");
    }

    public async Task<byte[]> GetConsentSignatureBytesAsync()
    {
        var module = await _moduleTask.Value;
        return await module.InvokeAsync<byte[]>("getConsentSignatureBytes");
    }

    public async Task RequestFullscreenAsync()
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("requestKioskFullscreen");
    }

    /// <summary>Stops the camera/MediaPipe/p5 instances between customers. The JS module stays loaded.</summary>
    public async Task TeardownAsync()
    {
        var module = await _moduleTask.Value;
        await module.InvokeVoidAsync("teardown");
    }

    public async ValueTask DisposeAsync()
    {
        if (_moduleTask.IsValueCreated)
        {
            var module = await _moduleTask.Value;
            await module.InvokeVoidAsync("teardown");
            await module.DisposeAsync();
        }
    }
}

File: VitalFace/src/VitalFace.Web/wwwroot/js/kiosk/kiosk-interop.js
// Single entry point Blazor imports as an ES module (see Services/KioskInterop.cs). Wires together
// the camera/shader visualization (p5-rppg-shader.js), the MediaPipe pipeline (face-bridge.js), the
// live metrics readout (pulse-viz.js) and the GDPR consent pad (signature-pad.js) behind a small,
// stable surface so the Razor component never touches the individual modules directly.

import { createFaceBridge } from './face-bridge.js';
import { createRppgAmplifier } from './p5-rppg-shader.js';
import { createPulseViz } from './pulse-viz.js';
import { createSignaturePad } from './signature-pad.js';

let amplifier = null;
let faceBridge = null;
let pulseViz = null;
let signaturePad = null;

export async function initializeCaptureStage(videoContainerId, vizContainerId) {
    amplifier = createRppgAmplifier(videoContainerId);
    const videoElement = await amplifier.ready();
    faceBridge = await createFaceBridge(videoElement);
    pulseViz = createPulseViz(vizContainerId);
}

export function startCapture() {
    faceBridge?.startCapture();
}

export function peekBuffers() {
    return faceBridge?.peekBuffers() ?? null;
}

export function stopCapture() {
    return faceBridge?.stopCapture() ?? { r: [], g: [], b: [], eyeOpenness: [], shoulderY: [], fps: 30 };
}

export function updateLiveMetrics(bpm, respPhase, perclos, confidence) {
    pulseViz?.update(bpm, respPhase, perclos, confidence);
}

export function initializeConsent(containerId) {
    signaturePad?.destroy();
    signaturePad = createSignaturePad(containerId);
}

export function clearSignature() {
    signaturePad?.clear();
}

export function hasConsentSignature() {
    return signaturePad?.hasSignature() ?? false;
}

export async function getConsentSignatureBytes() {
    return signaturePad ? await signaturePad.getPngBytes() : new Uint8Array();
}

export function requestKioskFullscreen() {
    const el = document.documentElement;
    if (el.requestFullscreen) {
        el.requestFullscreen().catch(() => { /* kiosk hardware may already be locked down at the OS level */ });
    }
}

// Tears down the camera, MediaPipe tasks and p5 sketches between customers (or on idle-timeout
// reset). The ES module itself stays loaded — initializeCaptureStage() rebuilds everything fresh
// for the next screening.
export function teardown() {
    faceBridge?.dispose();
    amplifier?.destroy();
    pulseViz?.destroy();
    signaturePad?.destroy();
    faceBridge = null;
    amplifier = null;
    pulseViz = null;
    signaturePad = null;
}
